plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/writer/GcsTransactionLogSynchronizer.java

```diff
@@ -0,0 +1,126 @@
+    }
+
+    // This approach should be compatible with OSS Delta Lake
+    // https://github.com/delta-io/delta/blob/master/storage/src/main/java/io/delta/storage/GCSLogStore.java
+    // This approach differs from the OSS Delta Lake corresponding implementation
+    // in the sense that it uses the `Storage` GCS API client directly (instead of Hadoop's `FSDataOutputStream`)
+    // in order to avoid leaked output streams in case of I/O exceptions occurring while uploading
+    // the blob content.
+    @Override
+    public void write(ConnectorSession session, String clusterId, Path newLogEntryPath, byte[] entryContents)
+    {
+        Storage storage = gcsStorageFactory.create(session);
+        try {
+            createStorageObjectExclusively(newLogEntryPath.toUri(), entryContents, storage);
+        }
+        catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+    }
+
+    @Override
+    public boolean isUnsafe()
+    {
+        return false;
+    }
+
+    /**
+     * This method uses {@link Storage} Cloud Storage API client for creating synchronously a storage object
+     * with the specified path and content.
+     * <p>
+     * This approach avoids dealing with `FSDataOutputStream` exposed by the method
+     * `org.apache.hadoop.fs.FileSystem#create(org.apache.hadoop.fs.Path)` and the intricacies of handling
+     * exceptions which may occur while writing the content to the output stream.
+     */
+    private static StorageObject createStorageObjectExclusively(URI blobPath, byte[] content, Storage storage)
+            throws IOException
+    {
+        CreateFileOptions createFileOptions = CreateFileOptions.DEFAULT_NO_OVERWRITE;
+        CreateObjectOptions createObjectOptions = objectOptionsFromFileOptions(createFileOptions);
+        StorageResourceId storageResourceId = StorageResourceId.fromUriPath(blobPath, false);
+
+        StorageObject object =
+                new StorageObject()
+                        .setContentEncoding(createObjectOptions.getContentEncoding())
+                        .setMetadata(encodeMetadata(createObjectOptions.getMetadata()))
+                        .setName(storageResourceId.getObjectName());
+
+        InputStream inputStream = new ByteArrayInputStream(content, 0, content.length);
+        Storage.Objects.Insert insert = storage.objects().insert(
+                storageResourceId.getBucketName(),
+                object,
+                new InputStreamContent(createObjectOptions.getContentType(), inputStream));
+        // By setting `ifGenerationMatch` setting to `0`, the creation of the blob will succeed only
+        // if there are no live versions of the object. When the blob already exists, the operation
+        // will fail with the exception message `412 Precondition Failed`
+        insert.setIfGenerationMatch(0L);
+        insert.getMediaHttpUploader().setDirectUploadEnabled(true);
+        insert.setName(storageResourceId.getObjectName());
```

Comment:
Name is also set above (`.setName(storageResourceId.getObjectName())`).
do both matter?

Reply:
> do both matter?
No.
https://github.com/GoogleCloudDataproc/hadoop-connectors/blob/v2.2.8/gcsio/src/main/java/com/google/api/services/storage/preprod/Storage.java#L8483-L8491
Just checked also the tests and this statement is not needed.